File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
int helper(vector<int>& nums,int idx , vector<int>& dp){
  if(idx ==1 ) return max(nums[0],nums[1]);
  if(idx == 0) return nums[0];
 if(dp[idx]!=-1)return dp[idx];
  int count = nums[idx] + helper(nums,idx-2,dp);
  int nocount = helper(nums,idx-1,dp);

  return dp[idx] = max(count,nocount);
}
    int rob(vector<int>& nums) {
        if(nums.size() == 1)return nums[0];
       // if(nums.size() == 2) return max(nums[0],nums[1]);
        vector<int>dp(nums.size()+1,-1);
        return helper(nums,nums.size()-1,dp);
        
    }
};

File: 0198-house-robber/0198-house-robber.java
class Solution {
   
   int help(int idx, int[] dp , int[] nums){
    if(idx == 0) return nums[0];
    if(idx < 0) return 0;

    if(dp[idx] != -1) return dp[idx];

    int pick = nums[idx] + help(idx-2,dp,nums);
    int npick = help(idx-1,dp,nums);

    return dp[idx] = Math.max(pick,npick);

   }
    
    public int rob(int[] nums) {
        if(nums.length == 0) return 0;
        if(nums.length == 1) return nums[0];
        int[] dp = new int[nums.length];
        Arrays.fill(dp,-1);
         int n = nums.length-1;
        help(n,dp,nums);
       
       // if(dp[n] == -1) return 0;
        return dp[n];
    }
}